### ISSUE TITLE
Fine-grained file permissions on Windows

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#13233]
So far, the only `File::Permissions` values available on Windows are `0o444`, `0o555`, `0o666`, and `0o777`, derived solely from whether a file has the read-only attribute and whether it is a directory. But Windows does support its own set of file permissions, and we should try to map them to `File::Permissions` as closely as possible.

First, it is indeed possible to deny read access to a file:

https://github.com/crystal-lang/crystal/blob/8a509b667f65a3a1a4b9d95afcc377c4db5b7f63/spec/std/file_spec.cr#L879-L885

It could be made to pass as follows:

```crystal
win_path = Process.quote_windows(path)
username = `whoami`.strip
File.touch(path)
`icacls #{win_path} /deny #{username}:(GR)`
expect_raises(File::AccessDeniedError) { File.read(path) }
`icacls #{win_path} /remove:d #{username}` # unnecessary with #9903?
```

Effectively, therefore, `File.chmod` without the read bit should deny the generic read rights, and similarly for write and execute rights, whereas `File.info` should also return them.

Second, a file or directory may be owned by either a user or a group, so it should technically be possible to obtain permissions like `0o700` or `0o770`. The three POSIX classes would likely be mapped to the effective rights for Creator Owner ID (`S-1-3-0`), Creator Group ID (`S-1-3-1`), and World (`S-1-1-0`). Changing the permissions might be a bit harder. Note that Windows does not truly have a concept of a user's primary group.

Finally, `File.copy` and `FileUtils.cp` should copy all the file permissions, including ones not exposed by `File::Permissions`.

**Comment by HertzDevil** [comment]
`File.new` and similar methods accept a `perm` argument that specifies the permissions to use when a file is created. It defaults to `rw-r--r--` on Unix-like systems (which has [its own issue](https://github.com/crystal-lang/crystal/issues/13455)), but on Windows the real default would be simply not passing anything to the `lpSecurityAttributes` parameter of `LibC.CreateFileW`, just as what we have always done, and letting automatic inheritance of DACLs do the job. So we have to make `perm` nilable, let each `Crystal::System::File` implementation figure out its own appropriate default, and then, if `perm` is not nil, set extra permissions on the created file's DACL:

```crystal
handle = LibC.CreateFileW(
  System.to_wstr(filename),
  access | LibC::ACCESS_MASK::WRITE_DAC,
  LibC::DEFAULT_SHARE_MODE, # UNIX semantics
  nil,
  disposition,
  attributes,
  LibC::HANDLE.null
)
if perm
  LibC.GetSecurityInfo(
    handle,
    LibC::SE_OBJECT_TYPE::FILE_OBJECT,
    LibC::OWNER_SECURITY_INFORMATION | LibC::GROUP_SECURITY_INFORMATION | LibC::DACL_SECURITY_INFORMATION,
    out owner_sid,
    out group_sid,
    out dacl,
    nil,
    out security_descriptor,
  )

  aces = [] of LibC::EXPLICIT_ACCESS_W

  world_bits = perm & 0o007
  LibC.ConvertStringSidToSidW(System.to_wstr("S-1-1-0"), out world_sid)
  LibC.BuildTrusteeWithSidW(out world_trustee, world_sid)
  aces << LibC::EXPLICIT_ACCESS_W.new(
    grfAccessPermissions: LibC::ACCESS_MASK::GENERIC_READ,
    grfAccessMode: LibC::ACCESS_MODE::SET_ACCESS, # replaces existing grant/deny info
    grfInheritance: 0,
    trustee: world_trustee,
  )
  aces << LibC::EXPLICIT_ACCESS_W.new(
    grfAccessPermissions: LibC::ACCESS_MASK::FILE_ALL_ACCESS & ~LibC::ACCESS_MASK::FILE_GENERIC_READ,
    grfAccessMode: LibC::ACCESS_MODE::DENY_ACCESS, # deny everything else
    grfInheritance: 0,
    trustee: world_trustee,
  )

  # ditto for owner bits (perhaps group bits too)

  LibC.SetEntriesInAclW(aces.size, aces, dacl, out new_dacl)
  LibC.SetSecurityInfo(handle, LibC::SE_OBJECT_TYPE::FILE_OBJECT, LibC::DACL_SECURITY_INFORMATION, nil, nil, new_dacl, nil)
end
```

Adding the world permissions alone looks like this:

![image](https://github.com/crystal-lang/crystal/assets/1361918/85b71523-15ec-43a3-89e6-c638855736a4)

![image](https://github.com/crystal-lang/crystal/assets/1361918/fe7dc872-6015-4fc7-91cb-b55f4d74928c)

![image](https://github.com/crystal-lang/crystal/assets/1361918/228c78f5-7ffe-4494-aece-e9500e3aa0bf)

Of course, if we unconditionally set all world permissions like this then we don't even need the security descriptor to inherit ACEs from parent directories.

Note that `FILE_ALL_ACCESS` is _not_ equal to simply merging the read / write / execute permissions; the former also contains the rights to delete the file, change permissions, and take ownership. Those are things that cannot be represented by `File::Permissions`. At this point I am wondering if a general overhaul to `File::Permissions` should happen first before trying to map it to Windows-specific behavior.

For directories we also have control over multiple inheritance settings, but most likely they will use whatever are Windows' sensible defaults and not be configurable via the standard library.

`File.chmod` would do a similar thing: read the file's DACL, add the permissions as above, and possibly remove any inherited ACEs. But this time the file might have been created outside Crystal with its own DACL, and we might not want to remove them.

**Comment by straight-shoota** [comment]
> At this point I am wondering if a general overhaul to `File::Permissions` should happen first before trying to map it to Windows-specific behavior.

Fun anecdote: `File::Permissions` was introduced as an abstraction in #5584 precisely to enable Windows support =)

I wouldn't go too deep about this. `File::Permissions` should be good enough to cover most typical use cases on most systems. And it's easy enough to comprehend without going into details of OS-specifics. Let's try to map the ideas of `File::Permissiones` to Windows semantics as good as possible. On POSIX we're also already only representing partial information because ACLs are not covered. So I'd think that we shouldn't do too much ACL stuff on Windows.
The fact is that it's impossible to get a full picture without getting into lots of details and platform-specific complications. If we accept that, we can stick to the most simple and comprehensible implementation that covers the bits that are reasonably close. But there's no need to go extra miles for mapping some complex details because we won't be able to fully represent them in a portable way. So less might even be more.

We can take inspiration from other languages. And that actually turns out to be not that much. Rust for example only exposes whether the file is writable in their portable API: https://doc.rust-lang.org/std/fs/struct.Permissions.html
In a similar manner, Golang only uses owner writeability in Windows: https://pkg.go.dev/os#Chmod

**Comment by HertzDevil** [comment]
I suppose we could hold off the general enhancement of `File::Permissions` even after Windows support is official. As long as the owner-write bit is consistently controllable it should be possible to write specs only based on that one bit
